client/beefy/src/import.rs

```diff
@@ -0,0 +1,197 @@
+use sp_blockchain::well_known_cache_keys;
+use sp_consensus::Error as ConsensusError;
+use sp_runtime::{
+	generic::{BlockId, OpaqueDigestItemId},
+	traits::{Block as BlockT, Header as HeaderT, NumberFor},
+	Justification,
+};
+
+use beefy_primitives::{
+	crypto::{AuthorityId, Signature},
+	BeefyApi, ConsensusLog, ValidatorSet, BEEFY_ENGINE_ID,
+};
+
+use crate::{
+	justification::decode_and_verify_justification, notification::BeefyJustificationSender,
+	Client as BeefyClient,
+};
+
+/// Checks the given header for a consensus digest signalling a beefy authority set change
+/// and extracts it.
+fn find_beefy_authority_set_change<B: BlockT>(
+	header: &B::Header,
+) -> Option<ValidatorSet<AuthorityId>> {
+	let id = OpaqueDigestItemId::Consensus(&BEEFY_ENGINE_ID);
+
+	let filter_log = |log: ConsensusLog<AuthorityId>| match log {
+		ConsensusLog::AuthoritiesChange(change) => Some(change),
+		_ => None,
+	};
+
+	// find the first consensus digest with the right ID which converts to
+	// the right kind of consensus log.
+	header.digest().convert_first(|l| l.try_to(id).and_then(filter_log))
+}
+
+/// A block-import handler for BEEFY.
+///
+/// This scans each imported block for BEEFY justifications and verifies them.
+/// Wraps a type `inner` that implements [`BlockImport`] and ultimately defers to it.
+pub struct BeefyBlockImport<Backend, Block: BlockT, Client, I> {
+	client: Arc<Client>,
+	inner: I,
+	justification_sender: BeefyJustificationSender<NumberFor<Block>, Signature>,
+	_phantom: PhantomData<Backend>,
+}
+
+impl<BE, Block: BlockT, Client, I: Clone> Clone for BeefyBlockImport<BE, Block, Client, I> {
+	fn clone(&self) -> Self {
+		BeefyBlockImport {
+			client: self.client.clone(),
+			inner: self.inner.clone(),
+			justification_sender: self.justification_sender.clone(),
+			_phantom: PhantomData,
+		}
+	}
+}
+
+#[async_trait::async_trait]
+impl<BE, Block: BlockT, Client, I> BlockImport<Block> for BeefyBlockImport<BE, Block, Client, I>
+where
+	BE: Backend<Block>,
+	I: BlockImport<
+			Block,
+			Error = ConsensusError,
+			Transaction = sp_api::TransactionFor<Client, Block>,
+		> + Send
+		+ Sync,
+	Client: BeefyClient<Block, BE>,
+	Client::Api: BeefyApi<Block>,
+	for<'a> &'a Client:
+		BlockImport<Block, Error = ConsensusError, Transaction = TransactionFor<Client, Block>>,
+	TransactionFor<Client, Block>: 'static,
+{
+	type Error = ConsensusError;
+	type Transaction = TransactionFor<Client, Block>;
+
+	async fn import_block(
+		&mut self,
+		block: BlockImportParams<Block, Self::Transaction>,
+		new_cache: HashMap<well_known_cache_keys::Id, Vec<u8>>,
+	) -> Result<ImportResult, Self::Error> {
+		let hash = block.post_hash();
+		let number = *block.header.number();
+		let justifications = block.justifications.clone();
+		// If this block contains a beefy authority set change then it must have a beefy
+		// justification
+		let change = find_beefy_authority_set_change::<Block>(&block.header);
```

Comment:
What is this assumption based on?

Reply:
This was based on the mandatory blocks requirement in the docs